### DISCUSSION TITLE
IOM member creation does not handle given local axes properly

### DISCUSSION THREAD

**Discussion by Rockambole** [idea-statica/ideastatica-public#440] [General Discussion]
Hello,

I'm making an IOM/BIM link between our FEM analysis system and IDEAStatica. I've successfully transferred geometry, properties and loads. I've followed IOM Github examples. One thing is odd when creating members:

1. I pass required data to a helper routine (including member local x,y,z axes, see below), which in turn creates a LineSegment3D object. 
2. When the flag transformCoordSystem=false (default), the member is created WITHOUT my xl,yl,zl axes. Geometry looks fine, but loads are un-balanced due to default xl,yl,zl axes defined during member creation.
3. When I set the flag transformCoordSystem=true, the members are created, but in a random fashion (with some members overlapping). Far from expected.
4. What is wrong here? 

Public Shared Function CreateLineSegment3D(ByVal model As OpenModel, ByVal startNode As String, ByVal endNode As String, ByVal vecX As Vector3D, ByVal vecY As Vector3D, ByVal vecZ As Vector3D, Optional ByVal transformCoordSystem As Boolean = False) As LineSegment3D

		Dim segment3D As LineSegment3D = New LineSegment3D()

		segment3D.Id = model.GetMaxId(segment3D) + 1
		segment3D.StartPoint = New ReferenceElement(model.Point3D.FirstOrDefault(Function(item) item.Name = startNode))
		Segment3D.EndPoint = New ReferenceElement(model.Point3D.FirstOrDefault(Function(item) item.Name = endNode))

		If transformCoordSystem Then
			Dim system As New CoordSystemByVector With {
				.VecX = vecX,
				.VecY = vecY,
				.VecZ = vecZ
			}
			segment3D.LocalCoordinateSystem = system
		End If

		Return segment3D

	End Function
 
Regards, O.S.

**Comment by nathanlukeIDEA** [comment]
Hi @Rockambole , 

It is to be expected that you get to this point and need to carefully consider LCS's and how to map them between the two software. It is generally not a trivial task.

These are our local co-ordinate system definitions.https://developer.ideastatica.com/docs/iom/iom_coordinate_systems.html

The transfer depends on a number of factors:

 

1. **Global axis direction** is it z-up or y-up. We base everything on Z up. If it is y it can be done, but the model will need to be transformed so everything is Z-up as required in IDEA.
2. **Default local co-ordinate system**. We use x along, z-up (right handed coord system). If this is the same for your program it is easiest to always transform the CoordSystem (set it directly as per your program). 
 If you do not explicitly set the LCS we will default it to a Z-up. For beams this is usually okay, but may cause some unwanted LCS's for columns  - in the example this is when we are transforming the co-ordinate system to make it go to x-global direction for columns.
3. S**ection definition  orientation**. Again we use z-up (right hand coordinate system). Check this is the same for your program if not you may need to do some transformations to the LCS and mirror the sections.
![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/1c8df58a-2f00-4492-a925-f505f609099d)
4. Once this is all correct it will depend on the **internal force sign convention** to get the forces correct.
 
The best way to debug all this is by importing your IOM file into Checkbot standalone and using IOM Import. You can view the whole structure and view the local coordinate systems.
![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/dc68c9bb-9449-4492-bc87-c0b14ca84752)

Cheers
Nathan
